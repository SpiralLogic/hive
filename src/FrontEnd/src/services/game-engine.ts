import { GameId, GameState, Move, PlayerId } from '../domain';
import { HexEngine } from '../domain/engine';

export default class GameEngine implements HexEngine {
  public playerId: PlayerId;
  public initialGame: Promise<GameState>;

  constructor(existingConfig?: { route: string; gameId: string; playerId: string }) {
    this.playerId = Number(existingConfig?.playerId) || 0;
    this.initialGame =
      existingConfig && existingConfig.route === 'game'
        ? this.getExistingGame(existingConfig.gameId)
        : this.getNewGame();
  }

  getExistingGame = async (gameId: GameId): Promise<GameState> => {
    const response = await fetch(`/api/game/${gameId}`, {
      method: `GET`,
      headers: this.requestHeaders,
    });

    return response.json();
  };

  getNewGame = async (): Promise<GameState> => {
    const response = await fetch(`/api/new`, {
      method: `POST`,
      headers: this.requestHeaders,
      body: JSON.stringify(''),
    });
    return response.json();
  };

  move = async (gameId: GameId, move: Move, useAi: boolean): Promise<GameState> => {
    const response = await fetch(`/api/move/${gameId}`, {
      method: 'POST',
      headers: this.requestHeaders,
      body: JSON.stringify(move),
    });
    if (useAi) {
      return (
        await fetch(`/api/ai-move/${gameId}/1`, {
          method: 'POST',
          headers: this.requestHeaders,
          body: JSON.stringify(move),
        })
      ).json();
    }
    return response.json();
  };

  private requestHeaders = {
    Accept: 'application/json',
    'Content-Type': 'application/json',
  };
}
