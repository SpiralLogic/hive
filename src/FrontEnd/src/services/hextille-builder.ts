import { Cell, Row } from '../domain';

export class HextilleBuilder {
  private readonly hexagons: Cell[];
  private readonly firstRow: number;
  private readonly height: number;
  private readonly firstColumn: number;
  private readonly width: number;

  constructor(hexagons: Cell[]) {
    this.hexagons = hexagons.sort((c1, c2) => c1.coords.r - c2.coords.r || c1.coords.q - c2.coords.q);
    [this.firstRow, this.height] = this.determineHeight(this.hexagons);
    [this.firstColumn, this.width] = this.determineWidth(this.hexagons);
  }

  private determineWidth = (cells: Cell[]): [number, number] => {
    const [min, max] = cells.reduce(([m1, m2], c) => [Math.min(m1, c.coords.q), Math.max(m2, c.coords.q)], [
      0,
      0,
    ]);
    return [min, max - min + 1];
  };

  private determineHeight = (cells: Cell[]): [number, number] => {
    const firstCell = cells[0]!;
    const lastCell = cells[cells.length - 1]!;
    const height = lastCell.coords.r - firstCell.coords.r + 1;
    return [firstCell.coords.r - 1, height + 2];
  };
  private createEmptyRow = (i: number): Row => {
    return {
      id: this.firstRow + i,
      hidden: i === 0 || i === this.height - 1,
      cells: Array.from(Array(this.width).keys(), (j: number) => ({
        coords: { q: this.firstColumn + j, r: this.firstRow + i },
        tiles: [],
        hidden: true,
      })),
    };
  };

  private createEmptyRows = () => Array.from(Array(this.height).keys(), this.createEmptyRow);

  createRows = (): Row[] =>
    this.hexagons.reduce((rows, cell) => {
      rows[cell.coords.r - this.firstRow]!.cells[cell.coords.q - this.firstColumn] = cell;
      return rows;
    }, this.createEmptyRows());
}
