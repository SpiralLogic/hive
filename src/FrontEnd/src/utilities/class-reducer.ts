import { useReducer } from 'preact/hooks';

const classReducer = (
  initialClasses: string,
  action: { type: 'add' | 'remove'; classes: string[] }
): string => {
  const classList = new Set(initialClasses.split(' '));
  switch (action.type) {
    case `add`:
      action.classes.forEach((c) => classList.add(c));
      break;
    case 'remove':
      action.classes.forEach((c) => classList.delete(c));
      break;
    default:
      break;
  }
  classList.delete('');
  return Array.from(classList).join(' ');
};

export const useClassReducer = (
  initialClasses: string
): [string, (action: { type: 'add' | 'remove'; classes: string[] }) => void] =>
  useReducer(classReducer, initialClasses);
