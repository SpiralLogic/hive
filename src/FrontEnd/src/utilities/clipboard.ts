const fallbackCopyTextToClipboard = (text: string) => {
  const currentFocus = document.activeElement as HTMLElement;
  const textArea = document.createElement('textarea');
  textArea.value = text;

  textArea.style.top = '0';
  textArea.style.left = '0';
  textArea.style.position = 'fixed';

  document.body.appendChild(textArea);
  textArea.focus();
  textArea.select();
  try {
    document.execCommand('copy');
    if (currentFocus && currentFocus.focus) currentFocus.focus();
    return true;
  } catch (e) {
    //  if (process.env.NODE_ENV !== 'production') console.error(e);
  } finally {
    document.body.removeChild(textArea);
  }
  return false;
};

export const shareGame = async (url: string) => {
  const opponentGame = {
    title: 'Hive board game',
    text: 'Join me in a game of the Hive!',
    url,
  };
  try {
    await navigator.share(opponentGame);
    return false;
  } catch {
    if (navigator.clipboard && navigator.clipboard.writeText) {
      try {
        await navigator.clipboard.writeText(opponentGame.url);
        return true;
      } catch {
        return fallbackCopyTextToClipboard(opponentGame.url);
      }
    }
  }
  return false;
};
