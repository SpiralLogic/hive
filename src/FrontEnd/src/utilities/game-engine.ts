import { GameConnection, GameHandlers, HexEngine, OpponentSelectionHandler } from '../domain/engine';
import { GameId, GameState, Move, Tile } from '../domain';
import { HubConnection, HubConnectionBuilder, HubConnectionState } from '@microsoft/signalr';

const requestHeaders = {
  Accept: 'application/json',
  'Content-Type': 'application/json',
};

const moveTile = async (gameId: GameId, move: Move): Promise<GameState> => {
  const response = await fetch(`/api/move/${gameId}`, {
    method: 'POST',
    headers: requestHeaders,
    body: JSON.stringify(move),
  });

  return await response.json();
};

const getExistingGame = async (gameId: GameId): Promise<GameState> => {
  const response = await fetch(`/api/game/${gameId}`, {
    method: `GET`,
    headers: requestHeaders,
  });

  return response.json();
};

const getNewGame = async (): Promise<GameState> => {
  const response = await fetch(`/api/new`, {
    method: `POST`,
    headers: requestHeaders,
    body: JSON.stringify(''),
  });
  return response.json();
};

const connectGame = (gameId: GameId, handlers: GameHandlers): GameConnection => {
  const getConnection = (gameId: GameId): HubConnection => {
    const hubUrl = `${window.location.protocol}//${window.location.host}/gamehub/${gameId}`;
    const connection = new HubConnectionBuilder().withUrl(hubUrl).withAutomaticReconnect().build();
    connection.on('ReceiveGameState', handlers.updateGameState);
    if (handlers.opponentSelection) connection.on('OpponentSelection', handlers.opponentSelection);
    connection.start().then();
    return connection;
  };

  const connection = getConnection(gameId);

  const sendSelection: OpponentSelectionHandler = (type: 'select' | 'deselect', tile: Tile) =>
    connection.state === HubConnectionState.Connected &&
    connection.invoke('SendSelection', type, tile).catch(function (err) {
      return console.error(err.toString());
    });

  const closeConnection = () => {
    connection.off('ReceiveGameState', handlers.updateGameState);
    return connection.stop();
  };

  if (process.env.NODE_ENV !== 'production') {
    connection.onreconnecting((error) => console.warn(`reconnecting to game ${gameId} .. ${error}`));
    connection.onclose((error) => console.info(`connection closed to game ${gameId} .. ${error}`));
    connection.onreconnected((error) => {
      window.location.reload();
      console.info(`reconnected to game ${gameId} .. ${error}`);
    });
  }

  return {
    getConnectionState: () => connection.state,
    sendSelection,
    closeConnection,
  };
};

const Engine: HexEngine = {
  getNewGame,
  getExistingGame,
  moveTile,
  connectGame: connectGame,
};

export default Engine;
