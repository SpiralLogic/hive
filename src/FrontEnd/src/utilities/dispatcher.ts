import { useEffect } from 'preact/hooks';
import { HiveDispatcher, HiveEventListener, HiveIntent } from '../services';

const hiveDispatcher = new HiveDispatcher();

export const useHiveDispatcher = (): HiveDispatcher => hiveDispatcher;

export const addHiveDispatchListener = <T extends HiveIntent>(
  type: T['type'],
  listener: HiveEventListener<T>
): void => {
  useEffect(() => {
    const emitter = useHiveDispatcher();
    return emitter.add<T>(type, listener);
  });
};

export const dispatchHiveEvent = hiveDispatcher.dispatch;
