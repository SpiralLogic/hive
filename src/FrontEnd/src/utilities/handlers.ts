import { GameState, MoveTile, Tile } from '../domain';
import { MoveEvent, TileEvent } from '../services';
import { OpponentConnectedHandler, OpponentSelectionHandler } from '../domain/engine';
import { dispatchHiveEvent, useHiveDispatcher } from './hooks';

export function handleDragOver(ev: { preventDefault: () => void }): boolean {
  ev.preventDefault();
  return false;
}

export function handleDrop(ev: { preventDefault: () => void }): boolean {
  ev.preventDefault();
  return false;
}

export const isEnterOrSpace = (event: KeyboardEvent): boolean => event.key === 'Enter' || event.key === ' ';

const focusNext = (target: HTMLElement, direction: -1 | 1) => {
  const allTabbable = Array.from(document.querySelectorAll('*[tabindex]:not(.name)'));
  let index = allTabbable.indexOf(target);
  if (index + direction < 0) index = allTabbable.length;
  (allTabbable[(index + direction) % allTabbable.length] as HTMLElement).focus();
};

export const handleKeyboardNav = (e: Pick<KeyboardEvent, 'key' | 'target'>): boolean => {
  if (e.target instanceof HTMLElement) {
    switch (e.key) {
      case 'ArrowDown':
      case 'ArrowRight':
        focusNext(e.target, 1);
        return true;
      case 'ArrowUp':
      case 'ArrowLeft':
        focusNext(e.target, -1);
        return true;
    }
  }
  return false;
};

export const opponentSelectionHandler: OpponentSelectionHandler = (type, tile) => {
  if (!tile) return;
  if (type === 'select') {
    dispatchHiveEvent({ type: 'tileSelect', tile });
  } else if (type === 'deselect') {
    dispatchHiveEvent({ type: 'tileDeselect', tile });
  }
};

export const opponentConnectedHandler: OpponentConnectedHandler = (type) => {
  if (type === 'connect') {
    dispatchHiveEvent({ type: 'opponentConnected' });
  } else if (type === 'disconnect') {
    dispatchHiveEvent({ type: 'opponentDisconnected' });
  }
};

export const attachServerHandlers = (
  sendSelection: (type: 'select' | 'deselect', tile: Tile) => void,
  gameState: GameState,
  updateGameState: (value: GameState) => void,
  moveTile: MoveTile
) => {
  const hiveDispatcher = useHiveDispatcher();

  const selectionChangeHandler = (event: TileEvent) => sendSelection('select', event.tile);
  const deselectionChangeHandler = (event: TileEvent) => sendSelection('deselect', event.tile);
  const moveHandler = async (event: MoveEvent) => {
    const gt = await moveTile(event.move);
    return updateGameState(gt);
  };

  hiveDispatcher.add<MoveEvent>('move', moveHandler);
  hiveDispatcher.add<TileEvent>('tileSelected', selectionChangeHandler);
  hiveDispatcher.add<TileEvent>('tileDeselected', deselectionChangeHandler);
  return () => {
    hiveDispatcher.remove<MoveEvent>('move', moveHandler);
    hiveDispatcher.remove<TileEvent>('tileSelected', selectionChangeHandler);
    hiveDispatcher.remove<TileEvent>('tileDeselected', deselectionChangeHandler);
  };
};
