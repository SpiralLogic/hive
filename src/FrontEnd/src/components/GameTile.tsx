import { FunctionComponent, h } from 'preact';
import { Tile as TileType } from '../domain';

import { TileAction } from '../services';
import { addHiveDispatchListener, dispatchHiveEvent, useClassReducer } from '../utilities/hooks';
import { handleDrop, handleKeyboardNav, isEnterOrSpace } from '../utilities/handlers';
import { useEffect, useState } from 'preact/hooks';
import Tile from './Tile';

const tileSelector = `[tabindex].tile`;
const cellSelector = `[tabindex][role="cell"]`;
const playerSelector = `[tabindex].name`;

const GameTile: FunctionComponent<TileType> = (tile: TileType) => {
  const { id, moves, creature, playerId } = tile;
  const [focus, setFocus] = useState(tileSelector);
  const [classes, setClassList] = useClassReducer(`player${playerId} hex`);

  const deselect = () => {
    if (!classes.includes('selected')) return;
    setClassList({ type: 'remove', classes: ['selected'] });
    dispatchHiveEvent({ type: 'tileDeselected', tile: tile });
  };

  const select = () => {
    if (classes.includes('selected')) return;
    dispatchHiveEvent({ type: 'tileClear' });
    setClassList({ type: 'add', classes: ['selected'] });
    dispatchHiveEvent({ type: 'tileSelected', tile: tile });
  };

  addHiveDispatchListener<TileAction>('tileSelect', (event: TileAction) => {
    if (event.tile.id === id) select();
  });

  addHiveDispatchListener<TileAction>('tileDeselect', (event: TileAction) => {
    if (event.tile.id === id) deselect();
  });

  addHiveDispatchListener('tileClear', () => {
    deselect();
    setFocus('');
  });

  const handleDragStart = (event: DragEvent) => {
    event.stopPropagation();
    select();
    setClassList({ type: 'add', classes: ['beforeDrag'] });
    setTimeout(() => setClassList({ type: 'remove', classes: ['beforeDrag'] }), 1);
  };
  const handleDragEnd = () => {
    dispatchHiveEvent({ type: 'tileClear' });
    dispatchHiveEvent({ type: 'tileDropped', tile: tile });
  };
  const handleClick = (event: MouseEvent) => {
    event.stopPropagation();
    !classes.includes('selected') ? select() : deselect();
  };
  const handleKeyDown = (e: KeyboardEvent) => {
    if (handleKeyboardNav(e) || !isEnterOrSpace(e)) return;
    e.stopPropagation();
    if (!classes.includes('selected')) {
      select();
      setFocus(cellSelector);
    } else {
      deselect();
    }
  };
  const handleMouseLeave = (event: { currentTarget: HTMLElement }) => event.currentTarget.blur();

  useEffect(() => {
    if (!focus) return;
    const focusElement =
      document.querySelector<HTMLElement>(focus) || document.querySelector<HTMLElement>(playerSelector);
    focusElement?.focus();
    setFocus('');
  }, [focus]);

  const attributes = {
    title: creature,
    class: classes ,
    draggable: !!moves.length,
    tabindex: moves.length ? 0 : undefined,
  };
  const handlers = moves.length
    ? {
        onclick: handleClick,
        ondragstart: handleDragStart,
        ondragend: handleDragEnd,
        onkeydown: handleKeyDown,
        onmouseleave: handleMouseLeave,
        ondrop: handleDrop,
      }
    : {
        ondrop: handleDrop,
      };
  return <Tile creature={creature} {...attributes} {...handlers} />;
};
GameTile.displayName = 'Tile';
export default GameTile;
