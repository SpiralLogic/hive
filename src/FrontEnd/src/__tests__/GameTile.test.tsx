import { act, fireEvent } from '@testing-library/preact';
import { h } from 'preact';
import { HiveEvent, TileAction, TileEvent } from '../services';
import { useHiveDispatcher } from '../utilities/dispatcher';
import GameTile from '../components/GameTile';
import { renderElement, simulateEvent } from './test-helpers';

describe('tile Tests', () => {
  const tileCanMove = {
    id: 1,
    playerId: 1,
    creature: 'ant',
    moves: [{ q: 1, r: 1 }],
  };
  const tileNoMove = { id: 2, playerId: 0, creature: 'fly', moves: [] };

  const createTileCanMove = () => {
    return renderElement(<GameTile currentPlayer={1} {...tileCanMove} />);
  };

  const createTileNoMove = () => {
    return renderElement(<GameTile currentPlayer={1} {...tileNoMove} />);
  };

  const createDispatcher = (): [TileEvent[], () => void] => {
    const tileEvents: TileEvent[] = [];
    const listener = (e: TileEvent) => tileEvents.push(e);

    const dispatcher = useHiveDispatcher();
    const cleanup = dispatcher.add<TileEvent>('tileSelected', listener);
    return [tileEvents, cleanup];
  };

  describe('tile render', () => {
    it('has creature', () => {
      expect(createTileNoMove().querySelectorAll('.creature')).toHaveLength(1);
      expect(createTileCanMove().querySelectorAll('.creature')).toHaveLength(1);
    });
  });

  describe('tile events', () => {
    it('click emits tile start event', () => {
      const [tileEvents, cleanup] = createDispatcher();
      fireEvent.click(createTileCanMove());
      const expectedEvent: TileEvent = {
        type: 'tileSelected',
        tile: tileCanMove,
      };
      expect(tileEvents).toEqual(expect.arrayContaining([expect.objectContaining(expectedEvent)]));
      cleanup();
    });

    it('enter emits tile start event', () => {
      const [tileEvents, cleanup] = createDispatcher();
      fireEvent.keyDown(createTileCanMove(), { key: 'Enter' });
      const expectedEvent: TileEvent = {
        type: 'tileSelected',
        tile: tileCanMove,
        fromEvent: false,
      };
      expect(tileEvents).toEqual(expect.arrayContaining([expect.objectContaining(expectedEvent)]));
      cleanup();
    });

    it('enter selects tile', () => {
      const tile = createTileCanMove();
      fireEvent.keyDown(tile, { key: 'Enter' });

      expect(tile).toHaveClass('selected');
    });

    it('second enter deselects tile', () => {
      const tile = createTileCanMove();
      fireEvent.keyDown(tile, { key: 'Enter' });
      fireEvent.keyDown(tile, { key: 'Enter' });

      expect(tile).not.toHaveClass('selected');
    });

    it('arrow keys use handler', () => {
      const tile = createTileCanMove();
      fireEvent.keyDown(tile, { key: 'ArrowDown' });

      expect(tile).toHaveFocus();
    });

    it('space emits tile start event', () => {
      const [tileEvents, cleanup] = createDispatcher();
      fireEvent.keyDown(createTileCanMove(), { key: ' ' });
      const expectedEvent: TileEvent = {
        type: 'tileSelected',
        tile: tileCanMove,
      };
      expect(tileEvents).toEqual(expect.arrayContaining([expect.objectContaining(expectedEvent)]));
      cleanup();
    });

    it('click deselects previous selected tile', () => {
      const tile = createTileCanMove();
      fireEvent.click(tile);

      expect(tile).toHaveClass('selected');
    });

    it('clicking same tile doesnt fire a tile start event', () => {
      const mock = jest.spyOn(useHiveDispatcher(), 'dispatch');
      const tile = createTileCanMove();
      fireEvent.click(tile);

      mock.mockClear();
      fireEvent.click(tile);

      expect(useHiveDispatcher().dispatch).not.toHaveBeenCalledWith(
        expect.objectContaining({ type: 'tileSelect' })
      );
    });

    it('mouseLeave activates blur', () => {
      const tile = createTileCanMove();
      tile.focus();
      fireEvent.mouseLeave(tile);
      expect(tile).not.toHaveFocus();
    });

    it('is draggable when there are available moves', () => {
      expect(createTileCanMove()).toHaveAttribute('draggable', 'true');
    });

    it('is *not* draggable when there are no moves available', () => {
      expect(createTileNoMove()).not.toHaveAttribute('draggable', 'false');
    });

    it('on drag emits start event', () => {
      const [tileEvents, cleanup] = createDispatcher();
      fireEvent.dragStart(createTileCanMove());
      const expectedEvent: TileEvent = {
        type: 'tileSelected',
        tile: tileCanMove,
      };
      expect(tileEvents).toEqual(expect.arrayContaining([expect.objectContaining(expectedEvent)]));
      cleanup();
    });

    it('on dragEnd emits end event', () => {
      const dropEvents: TileEvent[] = [];
      useHiveDispatcher().add<TileEvent>('tileDropped', (e) => dropEvents.push(e));
      fireEvent.dragEnd(createTileCanMove());
      const expectedEvent: HiveEvent = {
        type: 'tileDropped',
        tile: tileCanMove,
      };

      expect(dropEvents).toEqual(expect.arrayContaining([expect.objectContaining(expectedEvent)]));
    });

    it(`tile can be selected via action`, () => {
      const tile = createTileCanMove();
      act(() => {
        useHiveDispatcher().dispatch<TileAction>({ type: 'tileSelect', tile: tileCanMove });
      }).catch(() => {});
      expect(tile).toHaveClass('selected');
    });

    it(`an already selected tile doesnt fire a selected event when selected`, () => {
      const selectEvents: TileEvent[] = [];
      useHiveDispatcher().add<TileEvent>('tileSelected', (e) => selectEvents.push(e));
      createTileCanMove();
      act(() => {
        useHiveDispatcher().dispatch<TileAction>({ type: 'tileSelect', tile: tileCanMove });
      }).catch(() => {});
      act(() => {
        useHiveDispatcher().dispatch<TileAction>({ type: 'tileSelect', tile: tileCanMove });
      }).catch(() => {});
      expect(selectEvents).toHaveLength(1);
    });

    it(`an already deselected tile doesnt fire a deselected event when deselected`, () => {
      const deselectEvents: TileEvent[] = [];
      useHiveDispatcher().add<TileEvent>('tileDeselected', (e) => deselectEvents.push(e));
      createTileCanMove();
      act(() => {
        useHiveDispatcher().dispatch<TileAction>({ type: 'tileSelect', tile: tileCanMove });
      }).catch(() => {});
      act(() => {
        useHiveDispatcher().dispatch<TileAction>({ type: 'tileDeselect', tile: tileCanMove });
      }).catch(() => {});
      act(() => {
        useHiveDispatcher().dispatch<TileAction>({ type: 'tileDeselect', tile: tileCanMove });
      }).catch(() => {});
      expect(deselectEvents).toHaveLength(1);
    });

    it(`tile can be deselected via action`, () => {
      const tile = createTileCanMove();
      act(() => {
        useHiveDispatcher().dispatch<TileAction>({ type: 'tileSelect', tile: tileCanMove });
      }).catch(() => {});
      act(() => {
        useHiveDispatcher().dispatch<TileAction>({ type: 'tileDeselect', tile: tileCanMove });
      }).catch(() => {});
      expect(tile).not.toHaveClass('selected');
    });

    it(`tile is only selected on matching select events`, () => {
      const tile = createTileCanMove();
      useHiveDispatcher().dispatch<TileAction>({ type: 'tileSelect', tile: tileNoMove });

      expect(tile).not.toHaveClass('selected');
    });

    it(`tile is only deselected on matching select events`, () => {
      const deselectEvents: TileEvent[] = [];
      useHiveDispatcher().add<TileEvent>('tileDeselected', (e) => deselectEvents.push(e));
      useHiveDispatcher().dispatch<TileAction>({ type: 'tileSelect', tile: tileCanMove });
      useHiveDispatcher().dispatch<TileAction>({ type: 'tileDeselect', tile: tileNoMove });

      expect(deselectEvents).toHaveLength(0);
    });

    it('default on drop is prevented', () => {
      expect(simulateEvent(createTileCanMove(), 'drop')).toHaveBeenCalledWith();
      expect(simulateEvent(createTileNoMove(), 'drop')).toHaveBeenCalledWith();
    });
  });

  describe('tile Snapshot', () => {
    it('can move matches current snapshot', () => {
      expect(createTileCanMove()).toMatchSnapshot();
    });

    it('no moves matches current snapshot', () => {
      expect(createTileNoMove()).toMatchSnapshot();
    });

    it('when stacked matched snapshot', () => {
      expect(renderElement(<GameTile currentPlayer={0} {...tileNoMove} stacked={true} />)).toMatchSnapshot();
    });
  });
});
