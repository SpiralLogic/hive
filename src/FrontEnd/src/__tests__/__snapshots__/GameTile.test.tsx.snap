// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tile Tests tile Snapshot can move matches current snapshot 1`] = `
<div
  class="player1 hex tile ant"
  draggable="true"
  tabindex="0"
  title="ant"
>
  <svg
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <use
      href="#hex"
    />
    <use
      class="creature"
      href="#ant"
    />
  </svg>
</div>
`;

exports[`tile Tests tile Snapshot no moves matches current snapshot 1`] = `
<div
  class="player0 hex tile fly"
  title="fly"
>
  <svg
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <use
      href="#hex"
    />
    <use
      class="creature"
      href="#fly"
    />
  </svg>
</div>
`;

exports[`tile Tests tile Snapshot no moves matches current snapshot 2`] = `
<div
  class="player0 hex stacked tile fly"
  title="fly"
>
  <svg
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <use
      href="#hex"
    />
    <use
      class="creature"
      href="#fly"
    />
  </svg>
</div>
`;

exports[`tile Tests tile Snapshot when stacked matched snapshot 1`] = `
<div
  class="player0 hex stacked tile fly"
  title="fly"
>
  <svg
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <use
      href="#hex"
    />
    <use
      class="creature"
      href="#fly"
    />
  </svg>
</div>
`;
